From 6cbdd109499397010d08ea1538b96ef177244f20 Mon Sep 17 00:00:00 2001
From: Guido Guenther <agx@sigxcpu.org>
Date: Sun, 22 Jun 2008 20:17:45 +0200
Subject: [PATCH] add guid so calendars get nicer names on disk

use shortname for groups and users. The groupPrefix makes sure these
don't overlap
---
 twistedcaldav/directory/nss.py |   15 +++++++++------
 1 files changed, 9 insertions(+), 6 deletions(-)

diff --git a/twistedcaldav/directory/nss.py b/twistedcaldav/directory/nss.py
index 4fb4cfc..c7de423 100644
--- a/twistedcaldav/directory/nss.py
+++ b/twistedcaldav/directory/nss.py
@@ -31,8 +31,8 @@ Example caldavd.plist configuration:
     <dict>
       <key>realmName</key>
       <string>Test Realm</string>
-      <!-- Groups starting with this prefix are considered calendarserver
-           groups -->
+      <!-- Groups starting with this prefix are considered calendarserver groups - 
+           note that _no_ user must use this prefix to avoid guid collissions! -->
       <key>groupPrefix</key>
       <string>caldavd-</string>
       <!-- Don't treat user id's smaller than firstValidUid as calendarserver users -->
@@ -100,6 +100,8 @@ class NssDirectoryService(DirectoryService):
                  firstValidGid, lastValidGid):
         super(NssDirectoryService, self).__init__()
         self.nsswitch = NsSwitch()
+        if not groupPrefix:
+            raise ValueError("groupPrefix cannot be empty")
         self.groupPrefix = groupPrefix
         self.realmName = realmName
         self.first_valid_uid = firstValidUid
@@ -181,11 +183,11 @@ class NssDirectoryRecord(DirectoryRecord):
     """
     Nss Directory Record
     """
-    def __init__(self, service, recordType, shortName, fullName = None, calendarUserAddresses = set()):
+    def __init__(self, service, recordType, shortName, guid, fullName=None, calendarUserAddresses=set()):
         super(NssDirectoryRecord, self).__init__(
             service               = service,
             recordType            = recordType,
-            guid                  = None,
+            guid                  = guid,
             shortName             = shortName,
             fullName              = fullName,
             calendarUserAddresses = calendarUserAddresses,
@@ -202,7 +204,8 @@ class NssUserRecord(NssDirectoryRecord):
         calendarUserAddresses = set()
         if service.mailDomain:
             calendarUserAddresses.add("mailto:%s@%s" % (shortName, service.mailDomain))
-        super(NssUserRecord, self).__init__(service, recordType, shortName, fullName, calendarUserAddresses)
+        super(NssUserRecord, self).__init__(service, recordType, shortName, guid=shortName,
+                                            fullName=fullName, calendarUserAddresses=calendarUserAddresses)
 
     def verifyCredentials(self, credentials):
         # FIXME: plugin in PAM authentication here if you want too - kerberos works
@@ -223,7 +226,7 @@ class NssGroupRecord(NssDirectoryRecord):
     """
     def __init__(self, service, recordType, groupName, members=()):
         shortName  = groupName.replace(service.groupPrefix,'',1)
-        super(NssGroupRecord, self).__init__(service, recordType, shortName)
+        super(NssGroupRecord, self).__init__(service, recordType, shortName, guid=groupName)
         self._members = members
 
     def members(self):
-- 
1.5.5.4

